core/logger/audit/audit_logger.go

```diff
@@ -0,0 +1,215 @@
+// AuditLogger instance. If the environment variables are not set, the logger
+// is disabled and short circuits execution via enabled flag.
+func NewAuditLogger(logger logger.Logger) (AuditLogger, error) {
+	// Start parsing environment variables for audit logger
+	auditLogsURL := os.Getenv("AUDIT_LOGS_FORWARDER_URL")
+	if auditLogsURL == "" {
+		// Unset, return a disabled audit logger
+		logger.Info("No AUDIT_LOGS_FORWARDER_URL environment set, audit log events will not be captured")
+
+		return &AuditLoggerService{}, nil
+	}
+
+	env := "production"
+	if os.Getenv("CHAINLINK_DEV") == "true" {
+		env = "develop"
+	}
+	hostname, err := os.Hostname()
+	if err != nil {
+		return &AuditLoggerService{}, errors.Errorf("Audit Log initialization error - unable to get hostname: %s", err)
+	}
+
+	// Split and prepare optional service client headers from env variable
+	headers := []serviceHeader{}
+	headersEncoded := os.Getenv("AUDIT_LOGS_FORWARDER_HEADERS")
+	if headersEncoded != "" {
+		headerLines := strings.Split(headersEncoded, "\\")
+		for _, header := range headerLines {
+			keyValue := strings.Split(header, "||")
+			if len(keyValue) != 2 {
+				return &AuditLoggerService{}, errors.Errorf("Invalid AUDIT_LOGS_FORWARDER_HEADERS value, single pair split on || required, got: %s", keyValue)
+			}
+			headers = append(headers, serviceHeader{
+				header: keyValue[0],
+				value:  keyValue[1],
+			})
+		}
+	}
+
+	loggingChannel := make(chan WrappedAuditLog, AUDIT_LOGS_CAPACITY)
+
+	// Finally, create new auditLogger with parameters
+	auditLogger := AuditLoggerService{
+		logger:          logger.Helper(1),
+		enabled:         true,
+		serviceURL:      auditLogsURL,
+		serviceHeaders:  headers,
+		jsonWrapperKey:  os.Getenv("AUDIT_LOGS_FORWARDER_JSON_WRAPPER_KEY"),
+		environmentName: env,
+		hostname:        hostname,
+		localIP:         getLocalIP(),
+		loggingChannel:  loggingChannel,
+	}
+
+	// Start our go routine that will receive logs and send them out to the
+	// configured service.
+	go auditLogger.auditLoggerRoutine()
+
+	return &auditLogger, nil
+}
+
+// / Entrypoint for new audit logs. This buffers all logs that come in they will
+// / sent out by the goroutine that was started when the AuditLoggerService was
+// / created. If this service was not enabled, this immeidately returns.
+// /
+// / This function never blocks.
+func (l *AuditLoggerService) Audit(ctx context.Context, eventID EventID, data Data) {
+	if !l.enabled {
+		return
+	}
+
+	wrappedLog := WrappedAuditLog{
+		eventID: eventID,
+		data:    data,
+	}
+
+	select {
+	case l.loggingChannel <- wrappedLog:
+	default:
+		if len(l.loggingChannel) == AUDIT_LOGS_CAPACITY {
+			l.logger.Errorw("Audit log buffer is full. Dropping log with eventID: %s", eventID)
+		} else {
+			l.logger.Errorw("Could not send log to audit subsystem even though queue has %d space", AUDIT_LOGS_CAPACITY-len(l.loggingChannel))
+		}
```

Comment:
I think we can simplify this section by dropping the second case since it should never happen by definition.
```suggestion
		l.logger.Errorw("Audit log buffer is full. Dropping log with eventID: %s", eventID)
```

Reply:
I like having it because it handles all cases and in the event something unexpected happens we will have some indiction that something went wrong.

Reply:
But the case it is attempting to handle can not happen by definition in the language. IMHO it is a confusing distraction to any readers. :shrug: 

Regardless, as written it could flag false positives with misleading data due to the racy `len` calls. If the channel was not ready to receive we already know that it was full or nil. The only thing to do here is to check for nil, otherwise it was necessarily full.
```suggestion
		if l.loggingChannel != nil {
			l.logger.Errorw("Audit log buffer is full. Dropping log with eventID: %s", eventID)
		} else {
			l.logger.Errorw("Audit log buffer is nil")
		}
```